input.c

```diff
@@ -211,11 +211,16 @@ static int do_http_get(struct http_get *hg, const char *uri, int redirections)
 	hg->proxy = NULL;
 	hg->code = -1;
 	hg->fd = -1;
-	if (http_parse_uri(uri, &hg->uri))
+	hg->is_https = is_https_url(uri);
+	conn->write = hg->is_https ? &https_write : &socket_write;
+	conn->read = hg->is_https ? &https_read : &socket_read;
+
+	if (parse_uri(uri, &hg->uri))
```

Comment:
Consider passing a uri string to connection to be able to hide the uri parsing inside/behind connection. Now in many places you first parse the url and after create a connection.